### PR TITLE
planner/cascades: add transformation rule PushTopNDownOuterJoin

### PR DESCRIPTION
**Pull request by hsqlu**
### What problem does this PR solve? 

This PR adds Transformation rule PushTopNDownOuterJoin which is a part of TopN push down in cascades planner.
refer to: #13709 

### What is changed and how it works?

The logic is the same with`planner/core/rule_topn_push_down.go/(*LogicalJoin).pushDownTopN(*LogicalTopN).`

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: planner/cascades/transformation_rules.go [hunk 1]
```diff
@@ -936,6 +937,78 @@ func (r *MergeAdjacentProjection) OnTransform(old *memo.ExprIter) (newExprs []*m
...
+	for i := range topN.ByItems {
+		newTopN.ByItems[i] = topN.ByItems[i].Clone()
+	}
+	newTopNGroup := memo.NewGroupExpr(newTopN)
+	newTopNGroup.SetChildren(childGroup)
+	newChild := memo.NewGroupWithSchema(newTopNGroup, childGroup.Prop.Schema)
+	return newChild
+}
+
+// OnTransform implements Transformation interface.
```
Comment (francis0407) [inline planner/cascades/transformation_rules.go]: add some detailed comments to explain how this rule works? like:
```
This rule will transform `TopN->OuterJoin->(OuterChild, InnerChild)` to 
`TopN->OuterJoin->(TopN->OuterChild, InnerChild)`
```

File: planner/cascades/transformation_rules.go [hunk 2]
```diff
@@ -981,6 +982,80 @@ func (r *MergeAdjacentProjection) OnTransform(old *memo.ExprIter) (newExprs []*m
...
+		memo.EngineTiDBOnly,
+		memo.NewPattern(memo.OperandJoin, memo.EngineTiDBOnly),
+	)
+	return rule
+}
+
+// Match implements Transformation interface.
+// Use appliedRuleSet in GroupExpr to avoid re-apply rules.
+func (r *PushTopNDownOuterJoin) Match(expr *memo.ExprIter) bool {
+	return !expr.GetExpr().HasAppliedRule(r)
```
Comment (francis0407) [inline planner/cascades/transformation_rules.go]: Could we check the join type here?

File: planner/cascades/transformation_rules.go [hunk 3]
```diff
@@ -981,6 +982,89 @@ func (r *MergeAdjacentProjection) OnTransform(old *memo.ExprIter) (newExprs []*m
...
+	default:
+		return false
+	}
+}
+
+func pushTopNDownOuterJoinToChild(topN *plannercore.LogicalTopN, outerGroup *memo.Group, innerGroup *memo.Group) *memo.Group {
+	for _, by := range topN.ByItems {
+		cols := expression.ExtractColumns(by.Expr)
+		for _, col := range cols {
+			if innerGroup.Prop.Schema.Contains(col) {
```
Comment (lzmhhh123) [inline planner/cascades/transformation_rules.go]: How about using `!outerGroup.Prop.Schema.Contains(col)`? It looks more sensible.

File: planner/cascades/transformation_rules.go [hunk 4]
```diff
@@ -981,6 +982,89 @@ func (r *MergeAdjacentProjection) OnTransform(old *memo.ExprIter) (newExprs []*m
...
+
+// OnTransform implements Transformation interface.
+// This rule will transform `TopN->OuterJoin->(OuterChild, InnerChild)` to `TopN->OuterJoin->(TopN->OuterChild, InnerChild)`
+func (r *PushTopNDownOuterJoin) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	topN := old.GetExpr().ExprNode.(*plannercore.LogicalTopN)
+	joinExpr := old.Children[0].GetExpr()
+	join := joinExpr.ExprNode.(*plannercore.LogicalJoin)
+	joinSchema := old.Children[0].Group.Prop.Schema
+	leftGroup := old.Children[0].GetExpr().Children[0]
+	rightGroup := old.Children[0].GetExpr().Children[1]
```
Comment (lzmhhh123) [inline planner/cascades/transformation_rules.go]: %s/old.Children[0].GetExpr().Children[0]/joinExpr.Children[0]

**Comment by hsqlu** [comment]
> Would you like to add some integration tests for the rule?

How? Add some more cases to `TestJoin` in integration_suite_in.json?

File: planner/cascades/transformation_rules.go (hunk 3 above)
Comment (hsqlu) [reply planner/cascades/transformation_rules.go]: innerGroup contains the column doesn't mean the outerGroup doesn't. Are you sure about this change? I didn't think too much about that 😂

Comment (lzmhhh123) [reply planner/cascades/transformation_rules.go]: Yep, I think for all situations, checking `!outerGroup` is better than checking `innerGroup`. If you have some cases, we can have a deeper discussion.

**Comment by lzmhhh123** [comment]
> > Would you like to add some integration tests for the rule?
> 
> How? Add some more cases to `TestJoin` in integration_suite_in.json?

You can take a look at `integration_test.go` and the test data in directory "testdata".

File: planner/cascades/transformation_rules.go [hunk 5]
```diff
@@ -981,6 +982,89 @@ func (r *MergeAdjacentProjection) OnTransform(old *memo.ExprIter) (newExprs []*m
...
+	join := expr.Children[0].GetExpr().ExprNode.(*plannercore.LogicalJoin)
+	switch join.JoinType {
+	case plannercore.LeftOuterJoin, plannercore.LeftOuterSemiJoin, plannercore.AntiLeftOuterSemiJoin, plannercore.RightOuterJoin:
+		return true
+	default:
+		return false
+	}
+}
+
+func pushTopNDownOuterJoinToChild(topN *plannercore.LogicalTopN, outerGroup *memo.Group, innerGroup *memo.Group) *memo.Group {
```
Comment (zz-jason) [inline planner/cascades/transformation_rules.go]: how about bind this function to struct `PushTopNDownOuterJoin`?

**Comment by hsqlu** [comment]
>
> You can take a look at `integration_test.go` and the test data in directory "testdata".

After I added "select * from t1 join t2 on t1.a = t2.a order by t1.a limit 1" in TestJoin of integration_suite_in.json, the `go test --record` failed and error is:

```
PANIC: integration_test.go:198: testIntegrationSuite.TestJoin

... Panic: runtime error: index out of range [0] with length 0 (PC=0x4030E31)

/usr/local/go/src/runtime/panic.go:679
  in gopanic
/usr/local/go/src/runtime/panic.go:75
  in goPanicIndex
/Users/qiannan/go/src/github.com/pingcap/tidb/planner/core/resolve_indices.go:87
  in PhysicalHashJoin.ResolveIndices
/Users/qiannan/go/src/github.com/pingcap/tidb/planner/core/resolve_indices.go:549
  in basePhysicalPlan.ResolveIndices
/Users/qiannan/go/src/github.com/pingcap/tidb/planner/core/resolve_indices.go:409
  in PhysicalTopN.ResolveIndices
optimize.go:119
  in Optimizer.FindBestPlan
/Users/qiannan/go/src/github.com/pingcap/tidb/planner/optimize.go:151
  in optimize
/Users/qiannan/go/src/github.com/pingcap/tidb/planner/optimize.go:50
  in Optimize
/Users/qiannan/go/src/github.com/pingcap/tidb/planner/core/planbuilder.go:2682
  in PlanBuilder.buildExplain
/Users/qiannan/go/src/github.com/pingcap/tidb/planner/core/planbuilder.go:387
  in PlanBuilder.Build
/Users/qiannan/go/src/github.com/pingcap/tidb/planner/optimize.go:115
  in optimize
/Users/qiannan/go/src/github.com/pingcap/tidb/planner/optimize.go:50
  in Optimize
/Users/qiannan/go/src/github.com/pingcap/tidb/executor/compiler.go:61
  in Compiler.Compile
/Users/qiannan/go/src/github.com/pingcap/tidb/session/session.go:1110
  in session.execute
/Users/qiannan/go/src/github.com/pingcap/tidb/session/session.go:1064
  in session.Execute
/Users/qiannan/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:144
  in TestKit.Exec
/Users/qiannan/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:235
  in TestKit.MustQuery
integration_test.go:217
  in testIntegrationSuite.TestJoin.func1
/Users/qiannan/go/src/github.com/pingcap/tidb/util/testutil/testutil.go:247
  in TestData.OnRecord
integration_test.go:215
  in testIntegrationSuite.TestJoin
/usr/local/go/src/reflect/value.go:321
  in Value.Call
```

@lzmhhh123 What's wrong with that? Cheers

**Comment by winoros** [comment]
@hsqlu 
Looking into the stack, you'll find that `p.LeftJoinKeys[i] = lArg.(*expression.Column)`. This line causes TiDB to panic. You can take a look where initialize the `LeftJoinKeys` too see why it's empty.
I'll check it later whether we can remove this field to make your contributions more easily, not to meet such strange panic.

File: planner/cascades/transformation_rules.go (hunk 5 above)
Comment (hsqlu) [reply planner/cascades/transformation_rules.go]: I don't think it‘s quite different, any reason for that?

**Pull request merged**